### PR TITLE
LibWeb: Begin working towards implementing {ByteLength,Counting}QueuingStrategy

### PR DESCRIPTION
**Pull request by shannonbooth**
The `readonly attribute Function size` is still left to be implemented for these interfaces.

### CODE REVIEW

File: Userland/Libraries/LibWeb/Streams/CountQueuingStrategy.h [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+// https://streams.spec.whatwg.org/#countqueuingstrategy
+class CountQueuingStrategy final : public Bindings::PlatformObject {
+    WEB_PLATFORM_OBJECT(CountQueuingStrategy, Bindings::PlatformObject);
+
+public:
+    static WebIDL::ExceptionOr<JS::NonnullGCPtr<CountQueuingStrategy>> construct_impl(JS::Realm&, QueuingStrategyInit const&);
+
+    virtual ~CountQueuingStrategy() override;
+
+    // The highWaterMark getter steps are:
```
Comment (kennethmyhra) [inline Userland/Libraries/LibWeb/Streams/CountQueuingStrategy.h:26]: Wire up the spec link (https://streams.spec.whatwg.org/#cqs-high-water-mark) here, and move this comment inside the getter.

File: Userland/Libraries/LibWeb/Streams/CountQueuingStrategy.cpp [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+ */
+
+#include <LibWeb/Bindings/CountQueuingStrategyPrototype.h>
+#include <LibWeb/Bindings/Intrinsics.h>
+#include <LibWeb/Streams/CountQueuingStrategy.h>
+#include <LibWeb/WebIDL/ExceptionOr.h>
+
+namespace Web::Streams {
+
+WebIDL::ExceptionOr<JS::NonnullGCPtr<CountQueuingStrategy>> CountQueuingStrategy::construct_impl(JS::Realm& realm, QueuingStrategyInit const& init)
```
Comment (kennethmyhra) [inline Userland/Libraries/LibWeb/Streams/CountQueuingStrategy.cpp:14]: Add spec link (https://streams.spec.whatwg.org/#blqs-constructor)

File: Userland/Libraries/LibWeb/Streams/ByteLengthQueuingStrategy.cpp [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+ */
+
+#include <LibWeb/Bindings/ByteLengthQueuingStrategyPrototype.h>
+#include <LibWeb/Bindings/Intrinsics.h>
+#include <LibWeb/Streams/ByteLengthQueuingStrategy.h>
+#include <LibWeb/WebIDL/ExceptionOr.h>
+
+namespace Web::Streams {
+
+WebIDL::ExceptionOr<JS::NonnullGCPtr<ByteLengthQueuingStrategy>> ByteLengthQueuingStrategy::construct_impl(JS::Realm& realm, QueuingStrategyInit const& init)
```
Comment (kennethmyhra) [inline Userland/Libraries/LibWeb/Streams/ByteLengthQueuingStrategy.cpp:14]: Spec link (https://streams.spec.whatwg.org/#blqs-constructor)

File: Userland/Libraries/LibWeb/Streams/ByteLengthQueuingStrategy.h [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+// https://streams.spec.whatwg.org/#bytelengthqueuingstrategy
+class ByteLengthQueuingStrategy final : public Bindings::PlatformObject {
+    WEB_PLATFORM_OBJECT(ByteLengthQueuingStrategy, Bindings::PlatformObject);
+
+public:
+    static WebIDL::ExceptionOr<JS::NonnullGCPtr<ByteLengthQueuingStrategy>> construct_impl(JS::Realm&, QueuingStrategyInit const&);
+
+    virtual ~ByteLengthQueuingStrategy() override;
+
+    // The highWaterMark getter steps are:
```
Comment (kennethmyhra) [inline Userland/Libraries/LibWeb/Streams/ByteLengthQueuingStrategy.h:26]: Also here move this comment inside getter and add spec link (https://streams.spec.whatwg.org/#blqs-high-water-mark)

**Review by kennethmyhra — CHANGES_REQUESTED** [review]
Left some nit regards to spec links

File: Userland/Libraries/LibWeb/Streams/CountQueuingStrategy.h (hunk 1 above)
Comment (shannonbooth) [reply Userland/Libraries/LibWeb/Streams/CountQueuingStrategy.h]: Ah, thanks, closer to how other places does it. Applied this change, and all others suggested. Thanks for the quick review! I'm still trying to figure out the size steps, a little confused by the spec and how to implement that right now :D

**Review by kennethmyhra — APPROVED** [review]
LGTM!

File: Tests/LibWeb/Text/expected/Streams/QueuingStrategy-high-water-mark.txt [hunk 5]
```diff
@@ -0,0 +1,20 @@
+======================================
+CountQueuingStrategy
+======================================
+'{"highWaterMark":2}' => 2
+'{"highWaterMark":-1}' => -1
+{"badKey":-1}' => Exception raised of TypeError
+{}' => Exception raised of TypeError
```
Comment (kennethmyhra) [inline Tests/LibWeb/Text/expected/Streams/QueuingStrategy-high-water-mark.txt:7]: I am not too familiar with our text tests, but why are these missing the single quote at the beginning?

```suggestion
'{"badKey":-1}' => Exception raised of TypeError
'{}' => Exception raised of TypeError
```

File: Tests/LibWeb/Text/expected/Streams/QueuingStrategy-high-water-mark.txt [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+"string instead"' => Exception raised of TypeError
+'{"highWaterMark":"wrongType"}' => NaN
+'{"highWaterMark":{}}' => NaN
+======================================
+ByteLengthQueuingStrategy
+======================================
+'{"highWaterMark":2}' => 2
+'{"highWaterMark":-1}' => -1
+{"badKey":-1}' => Exception raised of TypeError
+{}' => Exception raised of TypeError
```
Comment (kennethmyhra) [inline Tests/LibWeb/Text/expected/Streams/QueuingStrategy-high-water-mark.txt:17]: Same here

```suggestion
'{"badKey":-1}' => Exception raised of TypeError
'{}' => Exception raised of TypeError
```

File: Tests/LibWeb/Text/expected/Streams/QueuingStrategy-high-water-mark.txt (hunk 5 above)
Comment (shannonbooth) [reply Tests/LibWeb/Text/expected/Streams/QueuingStrategy-high-water-mark.txt]: :facepalm: I missed the ' in the `println` in the test. Fixed, thanks!

**Review by kennethmyhra — APPROVED** [review]
I'd say this is ready. Let's get it in before #19458.

File: Userland/Libraries/LibWeb/Streams/ByteLengthQueuingStrategy.cpp [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+    , m_high_water_mark(high_water_mark)
+{
+}
+
+ByteLengthQueuingStrategy::~ByteLengthQueuingStrategy() = default;
+
+JS::ThrowCompletionOr<void> ByteLengthQueuingStrategy::initialize(JS::Realm& realm)
+{
+    MUST_OR_THROW_OOM(Base::initialize(realm));
+    set_prototype(&Bindings::ensure_web_prototype<Bindings::CountQueuingStrategyPrototype>(realm, "ByteLengthQueuingStrategy"));
```
Comment (mattco98) [inline Userland/Libraries/LibWeb/Streams/ByteLengthQueuingStrategy.cpp:33]:
```suggestion
    set_prototype(&Bindings::ensure_web_prototype<Bindings::ByteLengthQueuingStrategyPrototype>(realm, "ByteLengthQueuingStrategy"));
```

**Review by mattco98 — CHANGES_REQUESTED** [review]

File: Userland/Libraries/LibWeb/Streams/ByteLengthQueuingStrategy.cpp (hunk 7 above)
Comment (shannonbooth) [reply Userland/Libraries/LibWeb/Streams/ByteLengthQueuingStrategy.cpp]: whoops, thanks @mattco98 !

**Review by mattco98 — APPROVED** [review]

**Pull request merged**